chore(sdk): Add InputUri and, OutputUris placeholders in v2

chensun opened the pull request:
**Description of your changes:**
- Add InputUri, OutputUri in the forked v2.components module.
- Error out on invalid combination of i/o type and placeholder usage.
  - Use inputPath for parameter types should fail.
  - Use inputValue for artifact types should fail.
  - Use inputUri, outputUri for parameter types should fail.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->
- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).
<!--
    **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update before release.
-->

chensun commented:
/assign @numerology @neuromage 
/cc @Ark-kun

numerology commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+def _create_task_factory_from_component_spec(
+    component_spec: structures.ComponentSpec,
+    component_filename=None,
+    component_ref: structures.ComponentReference = None):
```
Okay to leave it as-is, so treat this comment as non-blocking:

component_ref will be eventually used as a k8s pod annotation when running in Argo, which is not yet implemented. We might not need this complexity here.

numerology commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+    input_path_generator: Callable[[str], str],
+    output_path_generator: Callable[[str], str],
+) -> _ResolvedCommandLineAndPaths:
+  """Resolves the command line argument placeholders.
```
optional: add a brief docstring documenting the usage/meaning of input/output_xxx_generator

numerology commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+  container_spec = component_spec.implementation.container
+
+  output_uris = OrderedDict()
+  output_paths = OrderedDict(
```
This line-break looks a bit of redundant.

numerology commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+
+  output_uris = OrderedDict()
+  output_paths = OrderedDict(
+  )  #Preserving the order to make the kubernetes output names deterministic
```
```suggestion
# Preserving the order to make the output names deterministic
```

numerology commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+        input_spec = inputs_dict[input_name]
+        if input_spec.optional:
+          #Even when we support default values there is no need to check for a default here.
+          #In current execution flow (called by python task factory), the missing argument would be replaced with the default value by python itself.
```
nit: line too long. Same for below

numerology commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+      from distutils.util import strtobool
+      condition_result_bool = condition_result and strtobool(
+          condition_result
+      )  #Python gotcha: bool('False') == True; Need to use strtobool; Also need to handle None and []
```
```suggestion
      )  # Python gotcha: bool('False') == True; Need to use strtobool; Also need to handle None and []
```

numerology commented on sdk/python/kfp/v2/components/python_op.py:
```diff
@@ -0,0 +1,1121 @@
+    'default_base_image_or_builder',
+    'InputUri',
+    'InputPath',
+    'InputTextFile',
```
Can we not expose `InputTextFile`/`InputBinaryFile` and their output counterparts for now before they're supported?

chensun replied:
IIUC, these could have been supported. They just translate into `{inputPath:}` placeholder.

numerology replied:
> They just translate into {inputPath:} placeholder.

This was a bit of surprising... Because when used with Python function component, `InputTextFile` actually indicates that system should pass a text stream object to the function, instead of a local path or URI.

Do you mind confirming what's the V1 behavior when used with ComponentSpec? Thanks!

numerology replied:
NVM, I just realized this is for Python function component, but not component spec. Lol

chensun replied:
Though they are for Python function component, I tried compiling such a function into yaml, and they resulted a component yaml definition with `String` typed input referenced with `{inputPath: }` placeholder. This violates our current rule that parameter input cannot be used with `{inputPath: }` placeholder. 

I'll remove these from the cleanup PR so that we don't expose unsupported features.

numerology reviewed (commented):
Thanks!

chensun commented on sdk/python/kfp/v2/components/components.py:
```diff
@@ -0,0 +1,667 @@
+
+    if isinstance(
+        argument_value,
+        (structures.GraphInputArgument, structures.TaskOutputArgument)):
```
We should have already throw errors on the pipeline level. One reason that so far I haven't tried touching this and applies to many other comments below was I wonder if this fork would be temporary and we might merge back to use the v1 Uri placeholder when ready.

chensun commented:
@numerology I feel the need to keep this PR as-is, while having the cleanup in a separate PR for cleaner and easier rollback or merge-back potential. WDYT?
Please take a look at 0ecc7cb from #4728. It didn't fix all the styling issues, but I believe it addressed most of your comments.

numerology commented on sdk/python/kfp/v2/dsl/container_op.py:
```diff
@@ -215,3 +217,23 @@ def set_gpu_limit(self, count: int) -> 'ContainerOp':
     return self
+
+  # Override _set_metadata to use v2 ComponentSpec
+  def _set_metadata(self, metadata):
```
Is it possible to do type hint here?
```suggestion
  def _set_metadata(self, metadata: structures.ComponentSpec):
```

chensun replied:
Done.

numerology commented on sdk/python/kfp/v2/dsl/container_op.py:
```diff
@@ -215,3 +217,23 @@ def set_gpu_limit(self, count: int) -> 'ContainerOp':
+      metadata (ComponentSpec): component metadata
+    """
+    if not isinstance(metadata, structures.ComponentSpec):
+      raise ValueError('_set_metadata is expecting ComponentSpec.')
```
I guess it should be TypeError, right?

chensun replied:
Fixed.

numerology reviewed (commented):
Thanks Chen!

numerology commented:
> @numerology I feel the need to keep this PR as-is, while having the cleanup in a separate PR for cleaner and easier rollback or merge-back potential. WDYT?

SGTM

numerology commented:
/lgtm
/approve

Thanks!
